datalad_ria/patches/sshremoteio.py

```diff
@@ -0,0 +1,83 @@
+"""
+
+This patch fixes the __init__ function of SSHRemoteIO, i.e.
+it adds error handling and a proper SSHRI-to-string function.
+"""
+
+import logging
+import subprocess
+
+from datalad import ssh_manager
+from datalad.support.exceptions import CommandError
+from datalad_next.patches import apply_patch
+
+# use same logger as -core
+lgr = logging.getLogger('datalad.customremotes.ria_remote')
+
+
+# This method is taken from
+# datalad@8a145bf432ae8931be7039c97ff602e53813d238
+# datalad/distributed/ore-remote.py:SSHRemoteIO.__init___
+
+
+DEFAULT_BUFFER_SIZE = 65536
+
+
+def SSHRemoteIO__init__(self, host, buffer_size=DEFAULT_BUFFER_SIZE):
```

Comment:
I would consider it good practice to add clear comment which part of the implementation are (or were) actually different from the original. At some point both will have evolved and we will not be able to tell easily how the are *meant* to differ.

Reply:
I think in principle it should be sufficient to state the complete method and the commit sha of the file. If the patch only introduced a single change, I would have marked that with a comment. In this case there are a few changes, and they are spread out. To quickly show the differences, I now included a listing of the original method into the comments (in 3222def56f1c28ca0476efad3236c342c6bb0963). WDYT?

Reply:
I was not saying that it is impossible to figure out what changed. But if I have only a code reference, I am forced to interpret the intention from various sources, and everyone can make there own conclusions and then we can discuss...

This is the full diff:

```diff
--- oldmethod.py        2023-09-19 12:37:10.612903880 +0200
+++ newmethod.txt       2023-09-19 12:37:28.715112889 +0200
@@ -1,10 +1,12 @@
-def __init__(self, host, buffer_size=DEFAULT_BUFFER_SIZE):
+def SSHRemoteIO__init__(self, host, buffer_size=DEFAULT_BUFFER_SIZE):
     """
     Parameters
     ----------
     host : str
       SSH-accessible host(name) to perform remote IO operations
       on.
+    buffer: int or None
+      The preferred buffer size
     """
 
     # the connection to the remote
@@ -14,8 +16,16 @@
         use_remote_annex_bundle=False,
     )
     self.ssh.open()
+
     # open a remote shell
-    cmd = ['ssh'] + self.ssh._ssh_args + [self.ssh.sshri.as_str()]
+    def copy_args(option):
+        args, open_args = self.ssh._ssh_args, self.ssh._ssh_open_args
+        if option in open_args and option not in args:
+            return [option, open_args[open_args.index(option) + 1]]
+        return []
+
+    copied_args = copy_args('-i') + copy_args('-p')
+    cmd = ['ssh'] + self.ssh._ssh_args + copied_args + [self.ssh.sshri.as_str()]
     self.shell = subprocess.Popen(cmd,
                                   stderr=subprocess.DEVNULL,
                                   stdout=subprocess.PIPE,
@@ -24,10 +34,15 @@
     self.shell.stdin.write(b"echo RIA-REMOTE-LOGIN-END\n")
     self.shell.stdin.flush()
     while True:
+        status = self.shell.poll()
+        if status not in (0, None):
+            raise CommandError(f'ssh shell process exited with {status}')
         line = self.shell.stdout.readline()
+        if line == '':
+            raise RuntimeError(f'ssh shell process close stdout unexpectedly')
         if line == b"RIA-REMOTE-LOGIN-END\n":
             break
     # TODO: Same for stderr?
 
-    # make sure default is used when None was passed, too.
+    # make sure default is used if 0 or None was passed, too.
     self.buffer_size = buffer_size if buffer_size else DEFAULT_BUFFER_SIZE
```

So apart from the doc-enhancement (non-critical), your patch changes how arguments are handled. This seems significant. I would need to play with it, to figure out what it does and not does. There is no comment.

What the error-handling code does is clear enough (for me), but that error handling was added at these points and only those, is all but obvious.

The full documentation of this patch at this point is

```
This patch fixes the __init__ function of SSHRemoteIO, i.e.
it adds error handling and a proper SSHRI-to-string function.
```

This is saying "if you have any question (and you will), come talk to me!" This is a feature of the code that we are trying to replace here. We cannot do the same mistake twice.

My point is, that we have to assume that the patch and the original implementation will evolve, in parallel and independently. We have to prepare for the situation that someone else will update the original implementation without  having seen this one. At that point, the patching will fail immediately. And then we need to act fast, and we should not require a research project that retraces the decision making.

It need not be as verbose as https://github.com/datalad/datalad-next/blob/main/datalad_next/patches/push_optimize.py

A simple: `# here is where the error handling is added` would go a long way already.

Reply:
I agree. Thanks for the explanation and the link. I updated the patch accordingly in d1d428a8f193d82198b89c986eafa55ac56e345d